lib/utils/prompt/context_reader.go

```diff
@@ -0,0 +1,282 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prompt
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/term"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ErrReaderClosed is returned from ContextReader.ReadContext after it is
+// closed.
+var ErrReaderClosed = errors.New("ContextReader has been closed")
+
+// ErrNotTerminal is returned by password reads attempted in non-terminal
+// readers.
+var ErrNotTerminal = errors.New("underlying reader is not a terminal")
+
+type readOutcome struct {
+	value []byte
+	err   error
+}
+
+type readerState int
+
+const (
+	readerStateIdle readerState = iota
+	readerStateClean
+	readerStatePassword
+	readerStateClosed
+)
+
+var (
+	termGetState     = term.GetState
+	termIsTerminal   = term.IsTerminal
+	termReadPassword = term.ReadPassword
+	termRestore      = term.Restore
```

Comment:
Could these be associated with a type, rather than being global variable functions?

Reply:
Sure. I'm not sure exactly what you had in mind, but here is my take at it.